test/tools/ossfuzz/YulOptimizerStepTest.cpp

```diff
@@ -0,0 +1,367 @@
+		ExpressionJoiner::run(*m_context, *m_ast);
+		ExpressionJoiner::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "loopInvariantCodeMotion")
+	{
+		disambiguate();
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+		FunctionHoister::run(*m_context, *m_ast);
+		LoopInvariantCodeMotion::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "controlFlowSimplifier")
+	{
+		disambiguate();
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+		ControlFlowSimplifier::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "structuralSimplifier")
+	{
+		disambiguate();
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+		FunctionHoister::run(*m_context, *m_ast);
+		LiteralRematerialiser::run(*m_context, *m_ast);
+		StructuralSimplifier::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "equivalentFunctionCombiner")
+	{
+		disambiguate();
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+		FunctionHoister::run(*m_context, *m_ast);
+		EquivalentFunctionCombiner::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "ssaReverser")
+	{
+		disambiguate();
+		SSAReverser::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "ssaAndBack")
+	{
+		disambiguate();
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+		// apply SSA
+		SSATransform::run(*m_context, *m_ast);
+		RedundantAssignEliminator::run(*m_context, *m_ast);
+		// reverse SSA
+		SSAReverser::run(*m_context, *m_ast);
+		FunctionHoister::run(*m_context, *m_ast);
+		CommonSubexpressionEliminator::run(*m_context, *m_ast);
+		UnusedPruner::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "stackCompressor")
+	{
+		disambiguate();
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+		FunctionHoister::run(*m_context, *m_ast);
+		FunctionGrouper::run(*m_context, *m_ast);
+		size_t maxIterations = 16;
+		StackCompressor::run(*m_dialect, *m_object, true, maxIterations);
+		BlockFlattener::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "wordSizeTransform")
+	{
+		disambiguate();
+		ExpressionSplitter::run(*m_context, *m_ast);
+		WordSizeTransform::run(*m_dialect, *m_dialect, *m_ast, *m_nameDispenser);
+	}
+	else if (m_optimizerStep == "fullSuite")
+	{
+		GasMeter meter(dynamic_cast<EVMDialect const&>(*m_dialect), false, 200);
+		OptimiserSuite::run(*m_dialect, &meter, *m_object, true, solidity::frontend::OptimiserSettings::DefaultYulOptimiserSteps);
+	}
+	else
+		solUnimplemented("Invalid optimization step");
+
+#if 0
+	std::cout << AsmPrinter{*m_dialect}(*m_ast) << std::endl;
+#endif
+
+	return m_ast;
+}
+
+void YulOptimizerStepTest::disambiguate()
+{
+	*m_ast = std::get<Block>(Disambiguator(*m_dialect, *m_analysisInfo)(*m_ast));
+	m_analysisInfo.reset();
+	updateContext();
+}
+
+void YulOptimizerStepTest::updateContext()
+{
+	m_nameDispenser = make_unique<NameDispenser>(*m_dialect, *m_ast, m_reservedIdentifiers);
+	m_context = make_unique<OptimiserStepContext>(OptimiserStepContext{
+		*m_dialect,
+		*m_nameDispenser,
+		m_reservedIdentifiers
+	});
+}
```

Comment:
This file has a lot of overlap with https://github.com/ethereum/solidity/blob/develop/test/libyul/YulOptimizerTest.cpp

The main difference is that for each optimiser step, I run all pre-requisite steps so that assertion failures due to not running them won't happen. For example for-loop-init-rewriter, if not run, can lead to failed assertions in other optimiser steps that depend on it.

Reply:
But is it the case that those optimizers are supposed to run it? I guess not. Do you still use `YulOptimizerTest.cpp`? Or is this file self sufficient?

Reply:
I'm just thinking if you couldn't just have a new class that takes care of the dependencies being correct, but still using `YulOptimizerTest`, since yea there's a lot of overlap

Reply:
To test whether an optimiser step by itself is correct, we will need to run all pre-requisites. The current testing framework (`YulOptimizerTest`) avoids running pre-reqs by test writers writing test cases that do not need these pre-requisites to be run. Example for loops, if present in test cases, are written such that they do not contain an init block etc.

But of course, a fuzzer does not understand optimiser steps and may synthesise for loops with an init block. All of this is fine as long as pre-requisites are run.

You are right. Perhaps, I should refactor the existing test framework such that pre-requisites may be run optionally, and run them in this PR

Reply:
I refactored existing yul optimised test framework so that portion of it could be reused by the fuzzer. If the checks in this PR pass, I will move these changes out into a smaller and more specific PR for ease of review.

Reply:
This PR now depends on #10699